File: engine/internal/modules/feedback/feedback_service.go
package feedback

import (
	"fmt"
	"net/http"
	"strconv"

	"github.com/Lands-Horizon-Corp/horizon-corp/internal/database/models"
	"github.com/Lands-Horizon-Corp/horizon-corp/internal/helpers"
	"github.com/Lands-Horizon-Corp/horizon-corp/internal/managers"
	"github.com/Lands-Horizon-Corp/horizon-corp/internal/providers"
	"github.com/Lands-Horizon-Corp/horizon-corp/server/middleware"
	"github.com/gin-gonic/gin"
)

type FeedbackService struct {
	controller    *managers.Controller[models.Feedback, models.FeedbackRequest, models.FeedbackResource]
	middle        *middleware.Middleware
	db            *providers.DatabaseService
	engine        *providers.EngineService
	models        *models.ModelResource
	tokenProvider *providers.TokenService
	helpers       *helpers.HelpersFunction
	modelResource *models.ModelResource
}

func NewFeedbackService(
	db *providers.DatabaseService,
	middle *middleware.Middleware,
	engine *providers.EngineService,
	models *models.ModelResource,
	tokenProvider *providers.TokenService,
	helpers *helpers.HelpersFunction,
	modelResource *models.ModelResource,
) *FeedbackService {
	controller := managers.NewController(
		models.FeedbackDB,
		models.ValidateFeedbackRequest,
		models.FeedbackToResource,
		models.FeedbackToResourceList,
	)

	return &FeedbackService{
		controller:    controller,
		middle:        middle,
		db:            db,
		engine:        engine,
		models:        models,
		tokenProvider: tokenProvider,
		helpers:       helpers,
		modelResource: modelResource,
	}
}

func (ts *FeedbackService) getUserClaims(ctx *gin.Context) (*providers.UserClaims, error) {
	claims, exists := ctx.Get("claims")
	if !exists {
		ts.tokenProvider.ClearTokenCookie(ctx)
		return nil, fmt.Errorf("claims not found in context")
	}
	userClaims, ok := claims.(*providers.UserClaims)
	if !ok {
		ts.tokenProvider.ClearTokenCookie(ctx)
		return nil, fmt.Errorf("failed to cast claims to *auth.UserClaims")
	}
	return userClaims, nil
}

// # Admin only
func (fs *FeedbackService) SearchFilter(ctx *gin.Context) {
	filterParam := ctx.Query("filter")
	if filterParam == "" {
		ctx.JSON(http.StatusBadRequest, gin.H{"error": "filter parameter is required"})
		return
	}
	pageIndexStr := ctx.Query("pageIndex")
	pageSizeStr := ctx.Query("pageSize")
	pageIndex, err := strconv.Atoi(pageIndexStr)
	if err != nil {
		pageIndex = 0
	}
	pageSize, err := strconv.Atoi(pageSizeStr)
	if err != nil {
		pageSize = 10
	}
	feedbacks, err := fs.modelResource.FeedbackFilterForAdmin(filterParam, pageIndex, pageSize)
	if err != nil {
		ctx.JSON(http.StatusInternalServerError, gin.H{"error": "Feedbacks not found."})
		return
	}
	data := fs.modelResource.FeedbackToResourceList(feedbacks.Data)
	if data == nil {
		ctx.JSON(http.StatusOK, gin.H{
			"data":      []interface{}{},
			"pageIndex": feedbacks.PageIndex,
			"totalPage": feedbacks.TotalPage,
			"pageSize":  feedbacks.PageSize,
			"totalSize": feedbacks.TotalSize,
			"pages":     feedbacks.Pages,
		})
		return
	}
	ctx.JSON(http.StatusOK, gin.H{
		"data":      data,
		"pageIndex": feedbacks.PageIndex,
		"totalPage": feedbacks.TotalPage,
		"pageSize":  feedbacks.PageSize,
		"totalSize": feedbacks.TotalSize,
		"pages":     feedbacks.Pages,
	})
}
func (fs *FeedbackService) ExportAll(ctx *gin.Context) {
	feedback, err := fs.modelResource.FeedbackDB.FindAll()
	if err != nil {
		ctx.String(http.StatusInternalServerError, "Failed to export CSV: %v", err)
		return
	}
	record, headers := fs.modelResource.FeedbackToRecord(feedback)
	csvManager := managers.NewCSVManager()
	csvManager.SetFileName("feedback-export-all.csv")
	csvManager.SetHeaders(headers)
	csvManager.AddRecords(record)
	if err := csvManager.WriteCSV(ctx); err != nil {
		ctx.String(http.StatusInternalServerError, "Failed to generate CSV: %v", err)
		return
	}
}

func (fs *FeedbackService) ExportAllFiltered(ctx *gin.Context) {
	filterParam := ctx.Query("filter")
	if filterParam == "" {
		ctx.JSON(http.StatusBadRequest, gin.H{"error": "filter parameter is required"})
		return
	}
	feedback, err := fs.modelResource.FeedbackFilterForAdminRecord(filterParam)
	if err != nil {
		ctx.JSON(http.StatusInternalServerError, gin.H{"error": "Companies not found."})
		return
	}
	record, headers := fs.modelResource.FeedbackToRecord(feedback)
	csvManager := managers.NewCSVManager()
	csvManager.SetFileName("feedback-export-all-filtered.csv")
	csvManager.SetHeaders(headers)
	csvManager.AddRecords(record)
	if err := csvManager.WriteCSV(ctx); err != nil {
		ctx.String(http.StatusInternalServerError, "Failed to generate CSV: %v", err)
		return
	}
}
func (fs *FeedbackService) ExportSelected(ctx *gin.Context) {
	ids, err := fs.helpers.ParseIDsFromQuery(ctx, "ids")
	if err != nil {
		ctx.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}
	feedbacks, err := fs.modelResource.FeedbackDB.GetAllByIDs(ids)
	if err != nil {
		ctx.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to retrieve companies."})
		return
	}
	record, headers := fs.modelResource.FeedbackToRecord(feedbacks)
	csvManager := managers.NewCSVManager()
	csvManager.SetFileName("feedback-export-all-filtered.csv")
	csvManager.SetHeaders(headers)
	csvManager.AddRecords(record)
	if err := csvManager.WriteCSV(ctx); err != nil {
		ctx.String(http.StatusInternalServerError, "Failed to generate CSV: %v", err)
		return
	}
}

func (as *FeedbackService) RegisterRoutes() {
	routes := as.engine.Client.Group("/api/v1/feedback")
	{
		// Public
		routes.POST("/", as.controller.Create)

		// Admin only
		routes.GET("/search", as.middle.AuthMiddlewareAdminOnly(), as.SearchFilter)
		routes.GET("/", as.middle.AuthMiddlewareAdminOnly(), as.controller.GetAll)
		routes.GET("/:id", as.middle.AuthMiddlewareAdminOnly(), as.controller.GetByID)
		routes.DELETE("/:id", as.middle.AuthMiddlewareAdminOnly(), as.controller.Delete)
		routes.DELETE("/bulk-delete", as.middle.AuthMiddlewareAdminOnly(), as.controller.DeleteMany)

		// Export routes
		routes.GET("/export", as.middle.AuthMiddlewareAdminOnly(), as.ExportAll)
		routes.GET("/export-search", as.middle.AuthMiddlewareAdminOnly(), as.ExportAllFiltered)
		routes.GET("/export-selected", as.middle.AuthMiddlewareAdminOnly(), as.ExportSelected)
	}
}

File: client/src/server/api-service/feedback-service.ts
import qs from 'query-string'

import APIService from './api-service'
import { downloadFile } from '../helpers'

import {
    TEntityId,
    IFeedbackRequest,
    IFeedbackResource,
    IFeedbackPaginatedResource,
} from '../types'

/**
 * Service class to handle CRUD operations for feedbacks.
 */
export default class FeedbackService {
    private static readonly BASE_ENDPOINT = '/feedback'

    public static async getById(
        id: TEntityId,
        preloads?: string[]
    ): Promise<IFeedbackResource> {
        const url = qs.stringifyUrl({
            url: `${FeedbackService.BASE_ENDPOINT}/${id}`,
            query: { preloads },
        })

        const response = await APIService.get<IFeedbackResource>(url)
        return response.data
    }

    public static async create(
        feedbackData: IFeedbackRequest
    ): Promise<IFeedbackResource> {
        const response = await APIService.post<
            IFeedbackRequest,
            IFeedbackResource
        >(FeedbackService.BASE_ENDPOINT, feedbackData)
        return response.data
    }

    public static async delete(id: TEntityId): Promise<void> {
        const endpoint = `${FeedbackService.BASE_ENDPOINT}/${id}`
        await APIService.delete<void>(endpoint)
    }

    public static async getFeedbacks(props?: {
        sort?: string
        filters?: string
        preloads?: string[]
        pagination?: { pageIndex: number; pageSize: number }
    }): Promise<IFeedbackPaginatedResource> {
        const { filters, preloads, pagination, sort } = props || {}

        const url = qs.stringifyUrl({
            url: `${FeedbackService.BASE_ENDPOINT}`,
            query: {
                sort,
                preloads,
                filter: filters,
                pageIndex: pagination?.pageIndex,
                pageSize: pagination?.pageSize,
            },
        })

        const response = await APIService.get<IFeedbackPaginatedResource>(url)
        return response.data
    }

    /**
     * Exports all feedbacks.
     *
     * @returns {Promise<void>} - A promise that resolves when the export is complete.
     */
    public static async exportAll(): Promise<void> {
        const url = `${FeedbackService.BASE_ENDPOINT}/export`
        await downloadFile(url, 'all_feedbacks_export.csv')
    }

    public static async exportAllFiltered(filters?: string): Promise<void> {
        const url = qs.stringifyUrl({
            url: `${FeedbackService.BASE_ENDPOINT}/export-search`,
            query: { filter: filters },
        })

        await downloadFile(url, 'filtered_feedbacks_export.csv')
    }

    public static async exportSelected(ids: TEntityId[]): Promise<void> {
        if (ids.length === 0) {
            throw new Error('No feedback IDs provided for export.')
        }

        const url = qs.stringifyUrl({
            url: `${FeedbackService.BASE_ENDPOINT}/export-selected`,
            query: { ids },
        })

        await downloadFile(url, 'selected_feedbacks_export.csv')
    }

    public static async deleteMany(ids: TEntityId[]): Promise<void> {
        const url = `${FeedbackService.BASE_ENDPOINT}/bulk-delete`
        const payload = { ids }

        await APIService.delete<void>(url, payload)
    }
}
